src/main/java/org/jenkinsci/plugins/configfiles/custom/security/CustomizedCredentialMapping.java

```diff
@@ -0,0 +1,107 @@
+ to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ copies of the Software, and to permit persons to whom the Software is
+ furnished to do so, subject to the following conditions:
+
+ The above copyright notice and this permission notice shall be included in
+ all copies or substantial portions of the Software.
+
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+ */
+package org.jenkinsci.plugins.configfiles.custom.security;
+
+import com.cloudbees.plugins.credentials.common.IdCredentials;
+import com.cloudbees.plugins.credentials.common.StandardListBoxModel;
+import com.cloudbees.plugins.credentials.common.StandardUsernameCredentials;
+import com.cloudbees.plugins.credentials.domains.DomainRequirement;
+import hudson.Extension;
+import hudson.model.AbstractDescribableImpl;
+import hudson.model.Descriptor;
+import hudson.model.Item;
+import hudson.model.ItemGroup;
+import hudson.model.Queue;
+import hudson.security.ACL;
+import hudson.security.AccessControlled;
+import hudson.util.ListBoxModel;
+import jenkins.model.Jenkins;
+import org.apache.commons.lang.StringUtils;
+import org.jenkinsci.plugins.plaincredentials.StringCredentials;
+import org.kohsuke.stapler.AncestorInPath;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.QueryParameter;
+
+import java.io.Serializable;
+import java.util.Collections;
+import java.util.List;
+
+public class CustomizedCredentialMapping extends AbstractDescribableImpl<CustomizedCredentialMapping> implements Serializable {
+
+    private final String tokenKey;
+    private final String credentialsId;
+
+    @DataBoundConstructor
+    public CustomizedCredentialMapping(String tokenKey, String credentialsId) {
+        this.tokenKey = tokenKey;
+        this.credentialsId = credentialsId;
+    }
+
+    public String getTokenKey() {
+        return tokenKey;
+    }
+
+    public String getCredentialsId() {
+        return credentialsId;
+    }
+
+    public Descriptor<CustomizedCredentialMapping> getDescriptor() {
+        return DESCRIPTOR;
+    }
+
+    private static final DescriptorImpl DESCRIPTOR = new DescriptorImpl();
+
+    @Extension
+    public static class DescriptorImpl extends Descriptor<CustomizedCredentialMapping> {
+
+        public ListBoxModel doFillCredentialsIdItems(@AncestorInPath ItemGroup context, @QueryParameter String tokenKey) {
+            AccessControlled _context = (context instanceof AccessControlled ? (AccessControlled) context : Jenkins.get());
```

Comment:
This isn't following the standard pattern: https://github.com/jenkinsci/credentials-plugin/blob/master/docs/consumer.adoc#providing-a-ui-form-element-to-let-a-user-select-credentials

Reply:
@jvz ah... ok. I lifted some of that code from elsewhere in the plugin.

The maven and properties credential mappings both do it that way as well. I was sort of feeling my way around trying to get this to work given my examples in the plugin only dealt with the StandardUsernameCredentials and I also need to add in the StringCredentials from plain-credentials plugin.

Reply:
Mentioning here as I said this in private: the code here isn't necessarily wrong, but it's overly restrictive in its permission checks. The checks for `Item.CONFIGURE` should be replaced by checks to `CredentialsProvider.USE_ITEM` when validating whether or not credentials should be allowed to be used or enumerated. This will allow Authorize Project to work properly here (otherwise, users with the ability to use credentials from a build as well as the permission to build the job would not be able to access the credentials when the job is run; the UseItem permission addresses this authorization issue). Essentially, a user who can build the job shouldn't also need permission to configure it just for the credentials to work (without Authorize Project installed, you wouldn't notice this problem since builds are executed as the system user normally, and that has permission to access almost any credentials).

I will note, though, that the `doFill...` and `doCheck...` methods for credentials can safely use `Item.CONFIGURE` as the permission check as that makes sense, but the code that resolves credentials ids into `IdCredentials` instances should certainly be checking the `CredentialsProvider.USE_ITEM` permission instead.